ddl/table.go

```diff
@@ -147,6 +148,146 @@ func onDropTableOrView(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	return ver, errors.Trace(err)
 }
 
+const (
+	restoreTableCheckFlagNone int64 = iota
+	restoreTableCheckFlagEnableGC
+	restoreTableCheckFlagDisableGC
+)
+
+func (w *worker) onRestoreTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	schemaID := job.SchemaID
+	tblInfo := &model.TableInfo{}
+	var autoID, dropJobID, restoreTableCheckFlag int64
+	var snapshotTS uint64
+	if err = job.DecodeArgs(tblInfo, &autoID, &dropJobID, &snapshotTS, &restoreTableCheckFlag); err != nil {
+		// Invalid arguments, cancel this job.
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	// check gc and safe point
+	gcEnable, err := checkGCEnable(w)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	err = checkTableNotExists(t, job, schemaID, tblInfo.Name.L)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Restore table divide into 2 steps:
+	// 1. Check gc enable status, to decided whether enable gc after restore table.
+	// 2. Do restore table job.
+	//     1. Check whether gc enabled, if enabled, disable gc first.
+	//     2. Check gc safe point. If drop table time if after safe point time, then can do restore.
+	//        otherwise, can't restore table, because the records of the table may already delete by gc.
+	//     3. Remove gc task of the table from gc_delete_range table.
+	//     4. Create table and rebase table auto ID.
+	//     5. Finish.
+	switch tblInfo.State {
+	case model.StateNone:
+		// none -> write only
+		// check gc enable and update flag.
+		if gcEnable {
+			job.Args[len(job.Args)-1] = restoreTableCheckFlagEnableGC
+		} else {
+			job.Args[len(job.Args)-1] = restoreTableCheckFlagDisableGC
+		}
+
+		job.SchemaState = model.StateWriteOnly
+		tblInfo.State = model.StateWriteOnly
+		ver, err = updateVersionAndTableInfo(t, job, tblInfo, false)
+	case model.StateWriteOnly:
+		// write only -> public
+		// do restore table.
+		if gcEnable {
+			err = disableGC(w)
+			if err != nil {
+				job.State = model.JobStateCancelled
+				return ver, errors.Errorf("disable gc failed, try again later. err: %v", err)
+			}
+		}
+		// check gc safe point
+		err = checkSafePoint(w, snapshotTS)
```

Comment:
Why not checkSafePoint in model.StateNone stage?

Reply:
I add a comment to explain this in #line 186.